docs/specs/buffer.md

```diff
@@ -0,0 +1,50 @@
+# Buffer Specification
+
+This document specifies Vector's buffer behavior for the development of Vector.
+
+The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”,
+“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be
+interpreted as described in [RFC 2119].
+
+<!-- MarkdownTOC autolink="true" style="ordered" indent="   " -->
+
+1. [Instrumentation](#instrumentation)
+
+<!-- /MarkdownTOC -->
+
+## Instrumentation
+
+Vector buffers MUST be instrumented for optimal observability and monitoring. This is required to drive various interfaces that Vector users depend on to manage Vector installations in mission critical production environments. This section extends the [Instrumentation Specification].
+
+### Events
+#### `EventsReceived`
+*All buffers* MUST emit an `EventsReceived` event immediately after receiving one or more Vector events.
+* Properties
+  * `entity` - buffer
+  * `count` - the number of received events
+  * `byte_size` - the byte size of received events
+* Metric
+  * MUST increment the `received_events_total` counter by the defined `count` with other properties as tags
+  * MUST increment the `received_event_bytes_total` counter by the defined `byte_size` with other properties as tags
+  * MUST increment the `buffer_event_count` gauge by the defined `count`
+  * MUST update the `buffer_usage_percentage` gauge which measures the current buffer space utilization (number of events/bytes) over total space available (max number of events/bytes)
```

Comment:
The `buffer_` prefix here makes sense, but feels out of line with other metrics. For example, `received_events_total` is shared across components and buffers, but this metric is scoped to buffers only. I think we have 1 of 2 paths:

1. Rename the metrics in the component spec to be prefixed with `component_ received_events_total`, etc.
2. Drop the `buffer_` prefix.

Reply:
Yeah, I agree, I think the `buffer_` "namespacing" is useful. Otherwise we need to differentiate with a metric label. Following that, adding a `component_` namespace to component specific metrics makes sense to me.